### ISSUE TITLE
How to convert handles to uint64_t?

### ISSUE THREAD

**Issue by haasn** [KhronosGroup/Vulkan-Docs#368]
At the very least, the `VK_EXT_debug_report` extension requires me to convert my handles to `uint64_t`, which doesn't seem to be sanely possible in a platform-neutral way.

For example, on 32-bit platforms, casting directly to `uint64_t` generates a compiler warning for dispatchable handles (-Wpointer-to-int-cast). You can silence this compiler warning by going through `uintptr_t` as well, i.e. `(uint64_t)(uintptr_t)handle`, but then you break non-dispatchable handles (which are 64-bits) since the cast through `uintptr_t` will slice off the upper 32 bits.

How am I expected to handle this? Right now, I have to define my own macro per handle type to do the correct conversion in a platform-neutral way. Could vulkan either be modified to make casting handles to `uint64_t` always safe, or otherwise provide macros for helping me deal with this?

**Comment by krOoze** [comment]
Also it uses some undefined `VK_NULL_OBJECT` instead of `VK_NULL_HANDLE`.

From the API point of view: what type is it supposed to accept, if it has to accept both dispatchable and non-dispatchable handles.

**Comment by courtney-g** [comment]
> VK_EXT_debug_report extension requires me to convert my handles

By this comment, you mean that the object handle in the callback is a uint64 and to test it against the application's handles requires this conversion, right?

The layers convert objects to uint64_t using reinterpret_cast<uint64_t &> object.
Some layers simply convert using (uint64_t), that should probably change to use reinterpret_cast above.

These do not appear to give us compiler warnings for 32-bit or 64-bit, but I'd have to dig deeper into the build to see what compiler options we use.

**Comment by haasn** [comment]
Sorry, I was thinking about VK_EXT_debug_marker, not VK_EXT_debug_report, but the issue should be the same in principle.

> These do not appear to give us compiler warnings for 32-bit or 64-bit, but I'd have to dig deeper into the build to see what compiler options we use.

Example of how to reproduce:

```
$ echo "int foo(void *t){ return (long long int)t; }" > test.c
$ gcc -std=c99 -Wall test.c -c -m32        
test.c: In function ‘foo’:
test.c:1:26: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
 int foo(void *t){ return (long long int)t; }
```

**Comment by courtney-g** [comment]
+baldurk
Baldur is the author of debug_marker and probably has a recommendation on what to do here.

**Comment by baldurk** [comment]
I don't have a strong recommendation, there doesn't seem like a clear 'best solution' winner. I don't think the API needs to change - debug_marker just inherited the uint64_t parameter from debug_report - the problem is how to make such a conversion acceptable to the compiler in all cases.

In C++ you could have an overloaded function that either takes void* or uint64_t, but C is more difficult. The 64-bit case can be handled with a simple macro, but as @haasn points out in 32-bit you *have* to use the uintptr_t cast for dispatchable handles, and *can't* use it for non-dispatchable.

I don't see a way to define a single macro that works for all handles on 32-bit in plain C. I don't think you can use the approach @krOoze suggested - the compiler will still warn on the second cast existing even if it can statically know that it won't be taken, even if you expand it to an `if {} else {}`. Short of just doing `memcpy(&uint, &handle, sizeof(handle))` I think you will need a separate macro for dispatchable or non-dispatchable handle casting.

My concern with this is that if someone uses the wrong one, they wouldn't get any compiler warnings but would truncate the non-dispatchable handle values. Vulkan does have a philosophy of requiring users to get things right, but having such macros in the official header encourages people to use them, and they're based on the most pessimistic/restricted use-case even when better solutions are available if the language/architecture/compiler is less restrictive. I'd lean towards saying this is something better to leave users to define per-project if it's needed based on what is available, unless someone has a clever macro trick that would let us define a single macro.

**Comment by ratchetfreak** [comment]
C11 has the [`_Generic`](http://en.cppreference.com/w/c/language/generic) selector. But I don't know how robust it will be.

**Comment by krOoze** [comment]
@courtney-g It applies to `VK_EXT_debug_report` too with `vkDebugReportMessageEXT` command, where you are supposed to supply `uint64_t object`.

@baldurk Right, should have tested it. Too much to hope for, that the compiler would recognize static if...

It is probably a non-problem (you can convert each handle depending on its type &mdash; and you do know the type). What makes it tricky is that the non-dispatchable handle can be a pointer or uint depending on platform.

**Comment by philiptaylor** [comment]
This appears to work for me in GCC 4.8, in 32-bit and 64-bit:

    #define HANDLE_TO_UINT64(h)                                 \
        __builtin_choose_expr(                                  \
            __builtin_types_compatible_p(typeof(h), uint64_t),  \
            (h),                                                \
            (uint64_t)(uintptr_t)(h)                            \
        )

which behaves kind of like a compile-time ternary expression.

I believe both branches of the `__builtin_choose_expr` still have to be valid warning-free expressions regardless of the condition. The trick is that the return type of the `__builtin_choose_expr` is determined from the type of the chosen branch, and the other branch will get ignored - in particular the other branch doesn't get converted to that return type, unlike a normal ternary expression, and that conversion is what would trigger the warning.

(This only works in C, not C++, so it needs an `#ifdef __cplusplus`. It does work in Clang 3.5, but in Clang you can do the simple `(uint64_t)(h)` instead and it doesn't seem to complain about the pointer-to-int cast.)

**Comment by TobiasHector** [comment]
@courtney-g @baldurk It looks like this has been sat for a while with no further action. In order for people to avoid bumping into this again, would one of you be willing to add an issue to the extension appendices for your extensions, documenting this weirdness and explaining, as best you can, how to deal with it?

**Comment by TobiasHector** [comment]
(Or possibly even better, in the function/struct descriptions that this affects)

**Comment by krOoze** [comment]
> This only works in C, not C++ [...]

There is a better way in C++ anyway. I actually was solving this way back for the purposes of the layers:
https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/b602f5d/layers/vk_layer_logging.h#L418

**Comment by courtney-g** [comment]
For the purpose of matching an application's handle with that returned by debug_report or debug_marker I recommend using the same casts that the validation layers use:
reinterpret_cast<uint64_t &>(dispatchableHandle)
(uint64_t)(nondispatchableHandle)

Unfortunately, when handles were broken into dispatchable and nondispatchable we were never able to come up with a universal method for comparing handles that worked for C, C++, 32bit and 64bit.

How's this as issue #4 for debug_report extension:
4) How do you compare handles returned by the debug_report callback to
the application's handles?

*RESOLVED*: Due to the different nature of dispatchable and nondispatchable
handles there is no generic way (that we know of) that works for common
compilers with 32bit, 64bit, C and C++. Thus recommend applications use
the same cast that the validation layers use:
	reinterpret_cast<uint64_t &>(dispatchableHandle)
	(uint64_t)(nondispatchableHandle)
This does require that the app treat dispatchable and nondispatchable handles
differently.

**Comment by timofonic** [comment]
@haasn @krOoze @oddhack @courtney-g @baldurk @ratchetfreak @philiptaylor @ghost Any news?

**Comment by oddhack** [comment]
This should be addressed in the 1.2.166 spec update.
